Documentation: Authorization (#3470)
neo4j/graphql: docs/modules/ROOT/pages/authentication-and-authorization/reference/operations.adoc

= Operations

This reference runs through a number of worked GraphQL queries and how each location in each query triggers the evaluation of different authentication/authorization rules.

> Each relevant location will have a comment such as `CREATE ON OBJECT Movie`, which means an authentication directive such as the following would be evaluated:

Reviewer: ```suggestion
Each relevant location has a comment such as `CREATE ON OBJECT Movie`, which means an authentication directive such as the following would be evaluated:
```

[source, graphql, indent=0]
----
type Movie @authentication(operations: [CREATE]) {
    title: String!
    actors: [Actor!]! @relationship(type: "ACTED_IN", direction: IN)
}
----

> This would also apply if the directive had no argumnts, because `operations` defaults to _all_ operations.

Reviewer: ```suggestion
This also applies if the directive had no arguments, because `operations` defaults to _all_ operations.
```

> The examples will apply to the `@authentication` directive, and also any rules within an `@authorization` directive.

Reviewer: ```suggestion
The following examples apply to the `@authentication` directive, and also any rules within an `@authorization` directive.
```

== Examples

=== Query

For a simple query, rules with `READ` in the operations will be evaluated for any type being read:

[source, graphql, indent=0]
----
query {
    movies {     # READ ON OBJECT Movie
        title    # READ ON FIELD_DEFINITION Movie.title
        actors { # READ ON OBJECT Actor
            name # READ ON FIELD_DEFINITION Actor.name
        }
    }
}
----

=== Mutation

> For create Mutations, `CREATE` rules on the object will be evaluated for each node created, as well as field definition rules:

Author: It's a bit scope creepy, but I've update this PR to include lower-casing all references to "Query", "Mutation" or "Subscription". 🙂
